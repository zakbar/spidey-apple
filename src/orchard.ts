export class Orchard {
  name: string;
  rate: number;
  isOpen: boolean;
  numberOfTree: number;
  towns: string[];
  weight: number;

  constructor(
    name: string,
    rate: number,
    isOpen: boolean,
    numberOfTree: number,
    towns: string[],
    weight: number = 0
  ) {
    this.name = name;
    this.rate = rate;
    this.isOpen = isOpen;
    this.numberOfTree = numberOfTree;
    this.towns = towns;
    this.weight = weight;
  }
  getWeight = () => {
    return this.weight;
  };

  close = () => {
    this.weight = 0;
    this.isOpen = false;
  };

  ensureClose = () => {
    if (this.numberOfTree <= this.rate) {
      this.close();
    }
  };
  commit = () => {
    if (this.isOpen) {
      this.numberOfTree -= this.rate;
      this.ensureClose();
    } else {
      this.numberOfTree += this.rate;
    }
    this.weight = 0;
  };

  open = () => {
    this.weight = 0.5;
    this.isOpen = true;
  };

  clone = () => {
    return new Orchard(
      this.name,
      this.rate,
      this.isOpen,
      this.numberOfTree,
      this.towns,
      this.weight
    );
  };

  canBeOpen = () => !this.isOpen && this.numberOfTree > this.rate;
}
