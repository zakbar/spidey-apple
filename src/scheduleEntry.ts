interface ScheduleEntry {
  name: string;
  isOpen: boolean;
  currentTree: number;
}
export { ScheduleEntry };
