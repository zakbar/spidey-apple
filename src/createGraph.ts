import { Graph } from "@dagrejs/graphlib";
import { Orchard } from "./orchard";

const createGraph = (orchards: Orchard[]): Graph => {
  const g = new Graph({ directed: false, multigraph: true });
  const towns = getUniqueTowns(orchards);
  for (const town of towns) {
    g.setNode(town);
  }
  for (const orchard of orchards) {
    g.setEdge(orchard.towns[0], orchard.towns[1], orchard.name);
  }
  return g;
};

const getUniqueTowns = (orchards: Orchard[]) => [
  ...new Set(orchards.flatMap((o) => o.towns)),
];

export { createGraph, getUniqueTowns };
