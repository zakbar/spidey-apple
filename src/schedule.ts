import { ScheduleEntry } from "./scheduleEntry";

export interface Schedule {
  orchardStatuses: ScheduleEntry[];
}
