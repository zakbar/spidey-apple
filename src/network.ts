import { alg, Edge, Graph } from "@dagrejs/graphlib";
import { Orchard } from "./orchard";
import { createGraph } from "./createGraph";

export interface Network {
  getSize: () => number;
  executeSearch: () => void;
  getOpenOrchards: () => Orchard[];
}

export class GraphLibNetwork implements Network {
  orchards: Orchard[];
  graph: Graph;
  openOrchards: Orchard[] = [];

  constructor(orchards: Orchard[]) {
    this.orchards = orchards;

    this.graph = createGraph(orchards);
  }

  getSize() {
    return this.graph.nodeCount();
  }
  executeSearch() {
    // Edge's prim algo weight evaluation is to check if it is best
    // to open in this period based on when was it open and
    // orchard's number of trees available against total available.
    // ref=https://en.wikipedia.org/wiki/Prim%27s_algorithm
    const mapWeight = new Map<string, number>();

    const totalOpen = this.orchards
      .filter((o) => o.isOpen)
      .reduce((partialSum, a) => partialSum + a.numberOfTree, 0);

    for (const orchard of this.orchards) {
      mapWeight.set(
        orchard.name,
        -(((1 + orchard.getWeight()) * orchard.numberOfTree) / totalOpen)
      );
    }

    const calculateWeight = (e: Edge) =>
      mapWeight.get(this.graph.edge(e.v, e.w) ?? "") ?? 1;

    const evaluatedGraphs = alg.prim(this.graph, calculateWeight);

    // End of prim algo calc

    const edges = evaluatedGraphs.edges();
    this.openOrchards = edges
      .map((e) =>
        this.orchards.find((o) => o.name === this.graph.edge(e.v, e.w))
      )
      .filter((o) => o !== undefined);
  }
  getOpenOrchards(): Orchard[] {
    return this.openOrchards;
  }
  getCloseOrchards(): Orchard[] {
    return this.orchards.filter(
      (o) => this.openOrchards.find((oc) => o === oc) === undefined
    );
  }
}
