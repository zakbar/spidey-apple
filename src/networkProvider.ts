import { alg } from "@dagrejs/graphlib";
import { GraphLibNetwork, Network } from "./network";
import { Orchard } from "./orchard";
import { createGraph } from "./createGraph";

export class GraphlibNetworkProvider {
  orchards: Orchard[];
  constructor(orchards: Orchard[]) {
    this.orchards = orchards;
  }
  getNetworks(): Network[] {
    const g = createGraph(this.orchards);

    const toReturn: Network[] = [];

    // Find all possible networks based on provided nodes and edges
    // http://en.wikipedia.org/wiki/Connected_component_(graph_theory)
    const possibleNetworks = alg.components(g);

    for (const possibleNetwork of possibleNetworks) {
      const possibleOrchards = this.orchards.filter((o) => {
        return (
          possibleNetwork.find((pn) => pn === o.towns[0]) &&
          possibleNetwork.find((pn) => pn === o.towns[1])
        );
      });
      toReturn.push(new GraphLibNetwork(possibleOrchards));
    }

    return toReturn;
  }
}
