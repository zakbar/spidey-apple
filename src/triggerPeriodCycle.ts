import { getCombinations } from "./getCombinations";
import { GraphlibNetworkProvider } from "./networkProvider";
import { Orchard } from "./orchard";
import { Schedule } from "./schedule";
import { ScheduleEntry } from "./scheduleEntry";

const processSinglePeriod = (orchards: Orchard[]) => {
  for (const orchard of orchards) {
    orchard.ensureClose();
  }
  const readyOrchards = orchards.filter((o) => o.isOpen);
  const networkProvider = new GraphlibNetworkProvider(readyOrchards);
  const networks = networkProvider
    .getNetworks()
    .sort((a, b) => a.getSize() - b.getSize());
  let orchardStatuses: ScheduleEntry[] = [];

  let allNetworkOpenOrchards: Orchard[] = [];
  for (const network of networks) {
    network.executeSearch();
    const openOrchards = network.getOpenOrchards();
    allNetworkOpenOrchards = [...allNetworkOpenOrchards, ...openOrchards];
    orchardStatuses = [
      ...orchardStatuses,
      ...openOrchards.map((l) => ({
        name: l.name,
        isOpen: true,
        currentTree: l.numberOfTree,
      })),
    ];
  }

  const openOrchards = orchardStatuses.map((o) => o.name);
  for (const orchard of orchards) {
    if (!openOrchards.find((o) => o === orchard.name)) {
      orchard.close();
      orchardStatuses.push({
        isOpen: false,
        name: orchard.name,
        currentTree: orchard.numberOfTree,
      });
    }
    orchard.commit();
  }

  return {
    orchardStatuses: orchardStatuses.sort((a, b) =>
      a.name.localeCompare(b.name)
    ),
  };
};

const simulateDoublePeriodMovement = (
  originals: Orchard[],
  towns: string[],
  opennableOrchards: string[],
  isLast: boolean = false
): { orchardStatuses: ScheduleEntry[] } | undefined => {
  const orchards = originals.map((o) => o.clone());
  if ((opennableOrchards?.length ?? 0) > 0) {
    for (const opennableOrchard of opennableOrchards!) {
      orchards.find((o) => o.name === opennableOrchard)?.open();
    }
  }
  let result = processSinglePeriod(orchards);

  if (!isLast) {
    const combination = combinationOfOpennableOrchards(orchards);

    while (combination.length > 0) {
      const possibleOpennableOrchards = combination.pop();
      if (possibleOpennableOrchards) {
        const result = simulateDoublePeriodMovement(
          orchards,
          towns,
          possibleOpennableOrchards,
          true
        );
        if (
          result &&
          isAllTownsServiced(towns, orchards, result.orchardStatuses)
        ) {
          return result;
        }
      }
    }
  } else {
    return result;
  }
};

const combinationOfOpennableOrchards = (orchards: Orchard[]) =>
  getCombinations(
    orchards
      .filter((o) => o.canBeOpen())
      .sort((a, b) => a.numberOfTree - b.numberOfTree)
      .map((o) => o.name)
  );

const getServicedTown = (orchards: Orchard[], openOrchards: ScheduleEntry[]) =>
  [
    ...new Set(
      orchards
        .filter((o) => openOrchards.find((f) => f.name === o.name && f.isOpen))
        .flatMap((o) => o.towns)
    ),
  ].sort((a, b) => a.localeCompare(b));

const isAllTownsServiced = (
  towns: string[],
  orchards: Orchard[],
  openOrchards: ScheduleEntry[]
) => {
  const servicedTowns = getServicedTown(orchards, openOrchards);
  const result = !towns.some((o) => !servicedTowns.find((t) => o === t));

  return result;
};

const triggerPeriodCycle = (
  originals: Orchard[],
  towns: string[]
): Schedule => {
  const opennableOrchardCombinations =
    combinationOfOpennableOrchards(originals);

  while (opennableOrchardCombinations.length > 0) {
    // Try to open closed orchards and simulate the next 2 periods to ensure all towns can be serviced.
    // Otherwise just pick whatever available.
    const opennableOrchards = opennableOrchardCombinations.pop();
    if (opennableOrchards && opennableOrchards.length > 0) {
      const nextPeriodResult = simulateDoublePeriodMovement(
        originals,
        towns,
        opennableOrchards
      );
      if (
        nextPeriodResult &&
        isAllTownsServiced(towns, originals, nextPeriodResult.orchardStatuses)
      ) {
        for (const opennableOrchard of opennableOrchards!) {
          originals.find((o) => o.name === opennableOrchard)?.open();
        }
        break;
      }
    }
  }
  return processSinglePeriod(originals);
};

export { triggerPeriodCycle, getServicedTown };
