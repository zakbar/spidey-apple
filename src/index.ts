import { Orchard } from "./orchard";
import { Schedule } from "./schedule";
import { getServicedTown, triggerPeriodCycle } from "./triggerPeriodCycle";

const townA = "A";
const townB = "B";
const townC = "C";
const townD = "D";
const townE = "E";
const towns = [townA, townB, townC, townD, townE];

const rate = 100;
const start = 501;
const numberOfPeriods = 12;
const orchard1 = new Orchard("1", rate, true, start, [townA, townB]);
const orchard2 = new Orchard("2", rate, true, start, [townA, townC]);
const orchard3 = new Orchard("3", rate, true, start, [townB, townC]);
const orchard4 = new Orchard("4", rate, true, start, [townB, townD]);
const orchard5 = new Orchard("5", rate, true, start, [townD, townE]);
const orchard6 = new Orchard("6", rate, true, start, [townC, townE]);
const orchards = [orchard1, orchard2, orchard3, orchard4, orchard5, orchard6];
const schedules: Schedule[] = [];

for (let i = 1; i <= numberOfPeriods; i++) {
  const schedule = triggerPeriodCycle(orchards, towns);
  schedules.push(schedule);
}

console.table(
  schedules.map((o, i) => {
    const obj: any = {};

    obj["period"] = i + 1;
    let total = 0;
    for (const status of o.orchardStatuses) {
      obj[`orchard ${status.name} status`] = status.isOpen ? "open" : "close";
      obj[`orchard ${status.name} # tree`] = status.currentTree;
      total += status.isOpen ? status.currentTree : 0;
    }
    obj["serviced towns"] = getServicedTown(orchards, o.orchardStatuses).join(
      ", "
    );
    obj["total tree"] = total;
    return obj;
  })
);
