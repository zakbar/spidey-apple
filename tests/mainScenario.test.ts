import { Orchard } from "../src/orchard";
import { Schedule } from "../src/schedule";
import { triggerPeriodCycle } from "../src/triggerPeriodCycle";

describe("12 Periods Scenario", () => {
  const townA = "A";
  const townB = "B";
  const townC = "C";
  const townD = "D";
  const townE = "E";
  const towns = [townA, townB, townC, townD, townE];

  const rate = 100;
  const start = 500;
  const numberOfPeriods = 12;
  const orchard1 = new Orchard("Orchard 1", rate, true, start, [townA, townB]);
  const orchard2 = new Orchard("Orchard 2", rate, true, start, [townA, townC]);
  const orchard3 = new Orchard("Orchard 3", rate, true, start, [townB, townC]);
  const orchard4 = new Orchard("Orchard 4", rate, true, start, [townB, townD]);
  const orchard5 = new Orchard("Orchard 5", rate, true, start, [townD, townE]);
  const orchard6 = new Orchard("Orchard 6", rate, true, start, [townC, townE]);
  const orchards = [orchard1, orchard2, orchard3, orchard4, orchard5, orchard6];
  const schedules: Schedule[] = [];

  for (let i = 1; i <= numberOfPeriods; i++) {
    test(`Period ${i}`, () => {
      const schedule = triggerPeriodCycle(orchards, towns);
      schedules.push(schedule);

      expect(schedule).toMatchSnapshot();
    });
  }
});
