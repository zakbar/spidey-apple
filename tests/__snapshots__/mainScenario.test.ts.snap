// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`12 Periods Scenario Period 1 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 500,
      "isOpen": false,
      "name": "Orchard 3",
    },
    {
      "currentTree": 500,
      "isOpen": false,
      "name": "Orchard 4",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 2 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 400,
      "isOpen": false,
      "name": "Orchard 1",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 600,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 600,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 400,
      "isOpen": false,
      "name": "Orchard 5",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 3 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 2",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 4 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 400,
      "isOpen": false,
      "name": "Orchard 3",
    },
    {
      "currentTree": 400,
      "isOpen": false,
      "name": "Orchard 4",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 5 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 1",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 500,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 5",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 6 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 2",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 7 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 3",
    },
    {
      "currentTree": 300,
      "isOpen": false,
      "name": "Orchard 4",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 8 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 1",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 400,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 5",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 9 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 2",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 10 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 2",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 4",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 5",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 11 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 1",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 2",
    },
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 3",
    },
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 4",
    },
    {
      "currentTree": 100,
      "isOpen": false,
      "name": "Orchard 5",
    },
    {
      "currentTree": 300,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;

exports[`12 Periods Scenario Period 12 1`] = `
{
  "orchardStatuses": [
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 1",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 2",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 3",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 4",
    },
    {
      "currentTree": 200,
      "isOpen": false,
      "name": "Orchard 5",
    },
    {
      "currentTree": 200,
      "isOpen": true,
      "name": "Orchard 6",
    },
  ],
}
`;
